README.md
<div align="center">

  <h1><code>gatsby-remark-embed-video-lite</code></h1>

[![npm](https://img.shields.io/npm/dt/gatsby-remark-embed-video-lite.svg?style=flat)](https://www.npmjs.com/package/gatsby-remark-embed-video-lite)

</div>

<div align="center">
<table width='100%' align="center">
    <tr>
        <td align='left' width='100%' colspan='2'>
            <strong><code>gatsby-remark-embed-video-lite</code></strong><br />
            📺 Embed a Youtube Video in your Markdown using high-performance web components
        </td>
    </tr>
</table>
</div>

<br>

Embed a Youtube or Vimeo video in your Markdown, using high-performance web components

Inspired by [gatsby-remark-embed-video](https://github.com/ahmadawais/gatsby-remark-better-embed-video)

## Install

1. Install plugin to your site:

```sh
npm i gatsby-remark-embed-video-lite
```

2. Add following to your `gatsby-config.js`:

```js
    plugins: [
      {
        resolve: "gatsby-transformer-remark",
        options: {
          plugins: [
          {
            resolve: "gatsby-remark-embed-video-lite",
          }
          ]
        }
      },
```

Uses the following components under the hood:

- https://www.npmjs.com/package/@justinribeiro/lite-youtube
- https://www.npmjs.com/package/@slightlyoff/lite-vimeo

3. Restart gatsby.

## Usage

```markdown
# Look at this Video:

`youtube: https://www.youtube.com/watch?v=2Xc9gXyf2G4`
`youtube: 2Xc9gXyf2G4`

`vimeo: https://vimeo.com/5299404`
`vimeo: 5299404`
```

## License

MIT


CHANGELOG.md
# [Unreleased](https://github.com/mattdsteele/gatsby-remark-embed-video-lite/compare/v0.1.1...23f495d3b03de6643bc71ea422d668cea58fc937) (2020-05-28)


### Bug Fixes

* update files to my liking ([23f495d](https://github.com/mattdsteele/gatsby-remark-embed-video-lite/commit/23f495d3b03de6643bc71ea422d668cea58fc937))



## [0.1.1](https://github.com/mattdsteele/gatsby-remark-embed-video-lite/compare/2.4.0...v0.1.1) (2020-05-27)



# [2.4.0](https://github.com/mattdsteele/gatsby-remark-embed-video-lite/compare/2.3.0...2.4.0) (2018-09-23)



# [2.3.0](https://github.com/mattdsteele/gatsby-remark-embed-video-lite/compare/2.2.0...2.3.0) (2018-09-23)



# [2.2.0](https://github.com/mattdsteele/gatsby-remark-embed-video-lite/compare/2.1.0...2.2.0) (2018-09-23)



# 2.1.0 (2018-09-23)





index.js
"use strict";
const EmbedVideo_1 = require("./EmbedVideo");
const util_1 = require("util");
const fs_1 = require("fs");
const visit = require(`unist-util-visit`);
const addVideoIframe = ({ markdownAST }, options) => {
    visit(markdownAST, `inlineCode`, (node) => {
        const { value } = node;
        let knownPlatforms = EmbedVideo_1.getKnownPlatforms();
        let keywords = [...knownPlatforms, 'video'].join('|');
        let re = new RegExp(`\(${keywords}\):\(\.\*\)`, 'i');
        const processValue = value.match(re);
        if (processValue) {
            let type = processValue[1];
            let id = processValue[2];
            id = id.trim();
            let embedVideo = new EmbedVideo_1.EmbedVideo(type, id, options);
            node.type = `html`;
            node.value = embedVideo.getHTML();
        }
    });
    const copy = util_1.promisify(fs_1.copyFile);
    // return Promise.all([
    //   copy(
    //     resolve(
    //       'node_modules/gatsby-remark-embed-video-lite/node_modules/@justinribeiro/lite-youtube/lite-youtube.js'
    //     ),
    //     'public/js/lite-youtube.js'
    //   ),
    //   copy(
    //     resolve(
    //       'node_modules/gatsby-remark-embed-video-lite/node_modules/@slightlyoff/lite-vimeo/lite-vimeo.js'
    //     ),
    //     'public/js/lite-vimeo.js'
    //   ),
    // ]);
};
module.exports = addVideoIframe;


index.ts
import { copyFile } from 'fs-extra';
import { EmbedVideo, EmbedVideoOptions, getKnownPlatforms } from './EmbedVideo';

import visit from 'unist-util-visit';

// Copy custom element files to static folder
const copyStaticFiles = () => {
  const ly = require.resolve('@justinribeiro/lite-youtube/lite-youtube.js');
  const lv = require.resolve('@slightlyoff/lite-vimeo/lite-vimeo.js');
  const DEPLOY_DIR = 'public/static';
  return Promise.all([
    copyFile(ly, `${DEPLOY_DIR}/lite-youtube.js`),
    copyFile(lv, `${DEPLOY_DIR}/lite-vimeo.js`),
  ]);
};

const addLiteElements = ({ markdownAST }: any, options: EmbedVideoOptions) => {
  visit(markdownAST, `inlineCode`, (node: { type: string; value: string }) => {
    const { value } = node;
    let knownPlatforms = getKnownPlatforms();
    let keywords = [...knownPlatforms, 'video'].join('|');
    let re = new RegExp(`\(${keywords}\):\(\.\*\)`, 'i');

    const processValue = value.match(re);
    if (processValue) {
      let type = processValue[1];
      let id = processValue[2];
      id = id.trim();

      let embedVideo = new EmbedVideo(type, id, options);

      node.type = `html`;
      node.value = embedVideo.getHTML();
    }
  });

  return copyStaticFiles();
};

export = addLiteElements;


EmbedVideo.ts
import { URL } from 'url';
export interface EmbedVideoOptions {
  width: number;
  ratio: number;
  related?: boolean;
  height?: number;
  noIframeBorder?: boolean;
  showInfo?: boolean;
}

interface VideoId {
  id: string;
  service: string;
}

const getVideoId = require('get-video-id');

const VideoServices: {
  [key: string]: string;
  YOUTUBE: string;
  VIMEO: string;
  VIDEOPRESS: string;
  TWITCH: string;
  TWITCHLIVE: string;
} = {
  YOUTUBE: 'youtube',
  VIMEO: 'vimeo',
  VIDEOPRESS: 'videopress',
  TWITCH: 'twitch',
  TWITCHLIVE: 'twitchlive',
};

export const getKnownPlatforms = () => {
  return Object.keys(VideoServices).map((val) => VideoServices[val] as string);
};

export class EmbedVideo {
  private options: EmbedVideoOptions;

  knownPlatforms = getKnownPlatforms();

  constructor(
    private type: string,
    private id: string,
    options?: EmbedVideoOptions
  ) {
    let defaultOptions = {
      width: 560,
      ratio: 1.77,
      related: false,
      noIframeBorder: false,
      showInfo: true,
    };
    this.options = { ...defaultOptions, ...options };

    if (!this.options.height) {
      this.options.height = Math.round(this.options.width / this.options.ratio);
    }
  }

  getHTML() {
    try {
      let videoId = this.readVideoId();
      let url = this.createUrl(videoId.service, videoId.id);
      let componentHtml = this.createComponent(videoId.service, url);
      return componentHtml;
    } catch (e) {
      return `<p style="color: red">Error: ${e.message}</p>`;
    }
  }

  private getTwitchId(input: string): VideoId | {} {
    let url;
    try {
      url = new URL(input);
      if (url.origin == 'https://player.twitch.tv') {
        let videoParam = url.searchParams.get('video');
        if (videoParam !== null) {
          return {
            id: videoParam,
            service: VideoServices.TWITCH,
          };
        }
        let channelParam = url.searchParams.get('channel');
        if (channelParam !== null) {
          return {
            id: channelParam,
            service: VideoServices.TWITCHLIVE,
          };
        }
      }

      if (url.origin == 'https://www.twitch.tv') {
        let pathSplit = url.pathname.split('/');
        if (pathSplit.length >= 2) {
          if (pathSplit[1] === 'videos') {
            if (pathSplit.length > 2) {
              return {
                id: `v${pathSplit[2]}`,
                service: VideoServices.TWITCH,
              };
            }
          } else {
            return {
              id: pathSplit[1],
              service: VideoServices.TWITCHLIVE,
            };
          }
        }
      }
    } catch (e) {
      return {};
    }

    return {};
  }

  private readVideoId(): VideoId {
    let videoId = getVideoId(this.id);

    if (videoId.id === undefined) {
      videoId = this.getTwitchId(this.id);
    }

    if (videoId.id === undefined) {
      if (this.type === 'video') {
        throw new TypeError('Id could not be processed');
      } else {
        return {
          id: this.id,
          service: this.type.toLowerCase(),
        };
      }
    }
    return videoId;
  }

  private createUrl(service: string, videoId: string): string {
    if (service === VideoServices.TWITCH && !videoId.startsWith('v')) {
      videoId = `v${videoId}`;
    }

    const urls: { [index: string]: string } = {
      youtube: `https://www.youtube.com/embed/${videoId}`,
      vimeo: `https://player.vimeo.com/video/${videoId}`,
      videopress: `https://videopress.com/embed/${videoId}`,
      twitch: `https://player.twitch.tv/?autoplay=false&video=${videoId}`,
      twitchlive: `https://player.twitch.tv/?channel=${videoId}`,
    };

    const url = new URL(urls[service]);

    if (!url) {
      throw new TypeError('Unknown Video Service');
    }

    const liteServices = [VideoServices.YOUTUBE, VideoServices.VIMEO];
    if (liteServices.includes(service)) {
      return videoId;
    }

    return url.toString();
  }

  private createComponent(videoPlatform: string, videoId: string) {
    switch (videoPlatform) {
      case VideoServices.YOUTUBE: {
        return ` <lite-youtube class="lite-video" videoid="${videoId}"></lite-youtube> `;
      }
      case VideoServices.VIMEO: {
        return `<lite-vimeo class="lite-video" videoid="${videoId}"></lite-vimeo>`;
      }
    }
    let iframeNode = `<iframe
              width="${this.options.width}"
              height="${this.options.height}"
              src="${videoId}"
              class="embedVideoIframe test-1"
              allowfullscreen
            ></iframe>`;
    if (this.options.noIframeBorder) {
      iframeNode += `
      <style>
        .embedVideoIframe {
          border: 0
        }
      </style>`;
    }

    if (videoPlatform === VideoServices.VIDEOPRESS) {
      iframeNode += `<script src="https://videopress.com/videopress-iframe.js"></script>`;
    }

    return iframeNode;
  }
}


EmbedVideo.js
"use strict";
/// <reference path="EmbedVideo.d.ts" />
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmbedVideo = exports.getKnownPlatforms = void 0;
const url_1 = require("url");
const getVideoId = require('get-video-id');
const VideoServices = {
    YOUTUBE: 'youtube',
    VIMEO: 'vimeo',
    VIDEOPRESS: 'videopress',
    TWITCH: 'twitch',
    TWITCHLIVE: 'twitchlive',
};
exports.getKnownPlatforms = () => {
    return Object.keys(VideoServices).map((val) => VideoServices[val]);
};
class EmbedVideo {
    constructor(type, id, options) {
        this.type = type;
        this.id = id;
        this.knownPlatforms = exports.getKnownPlatforms();
        let defaultOptions = {
            width: 560,
            ratio: 1.77,
            related: false,
            noIframeBorder: false,
            showInfo: true,
        };
        this.options = { ...defaultOptions, ...options };
        if (!this.options.height) {
            this.options.height = Math.round(this.options.width / this.options.ratio);
        }
    }
    getHTML() {
        try {
            let videoId = this.readVideoId();
            let url = this.createUrl(videoId.service, videoId.id);
            let componentHtml = this.createComponent(videoId.service, url);
            return componentHtml;
        }
        catch (e) {
            return `<p style="color: red">Error: ${e.message}</p>`;
        }
    }
    getTwitchId(input) {
        let url;
        try {
            url = new url_1.URL(input);
            if (url.origin == 'https://player.twitch.tv') {
                let videoParam = url.searchParams.get('video');
                if (videoParam !== null) {
                    return {
                        id: videoParam,
                        service: VideoServices.TWITCH,
                    };
                }
                let channelParam = url.searchParams.get('channel');
                if (channelParam !== null) {
                    return {
                        id: channelParam,
                        service: VideoServices.TWITCHLIVE,
                    };
                }
            }
            if (url.origin == 'https://www.twitch.tv') {
                let pathSplit = url.pathname.split('/');
                if (pathSplit.length >= 2) {
                    if (pathSplit[1] === 'videos') {
                        if (pathSplit.length > 2) {
                            return {
                                id: `v${pathSplit[2]}`,
                                service: VideoServices.TWITCH,
                            };
                        }
                    }
                    else {
                        return {
                            id: pathSplit[1],
                            service: VideoServices.TWITCHLIVE,
                        };
                    }
                }
            }
        }
        catch (e) {
            return {};
        }
        return {};
    }
    readVideoId() {
        let videoId = getVideoId(this.id);
        if (videoId.id === undefined) {
            videoId = this.getTwitchId(this.id);
        }
        if (videoId.id === undefined) {
            if (this.type === 'video') {
                throw new TypeError('Id could not be processed');
            }
            else {
                return {
                    id: this.id,
                    service: this.type.toLowerCase(),
                };
            }
        }
        return videoId;
    }
    createUrl(service, videoId) {
        if (service === VideoServices.TWITCH && !videoId.startsWith('v')) {
            videoId = `v${videoId}`;
        }
        const urls = {
            youtube: `https://www.youtube.com/embed/${videoId}`,
            vimeo: `https://player.vimeo.com/video/${videoId}`,
            videopress: `https://videopress.com/embed/${videoId}`,
            twitch: `https://player.twitch.tv/?autoplay=false&video=${videoId}`,
            twitchlive: `https://player.twitch.tv/?channel=${videoId}`,
        };
        const url = new url_1.URL(urls[service]);
        if (!url) {
            throw new TypeError('Unknown Video Service');
        }
        const liteServices = [VideoServices.YOUTUBE, VideoServices.VIMEO];
        if (liteServices.includes(service)) {
            return videoId;
        }
        return url.toString();
    }
    createComponent(videoPlatform, videoId) {
        switch (videoPlatform) {
            case VideoServices.YOUTUBE: {
                return ` <lite-youtube class="lite-video" videoid="${videoId}"></lite-youtube> `;
            }
            case VideoServices.VIMEO: {
                return `<lite-vimeo class="lite-video" videoid="${videoId}"></lite-vimeo>`;
            }
        }
        let iframeNode = `<iframe
              width="${this.options.width}"
              height="${this.options.height}"
              src="${videoId}"
              class="embedVideoIframe test-1"
              allowfullscreen
            ></iframe>`;
        if (this.options.noIframeBorder) {
            iframeNode += `
      <style>
        .embedVideoIframe {
          border: 0
        }
      </style>`;
        }
        if (videoPlatform === VideoServices.VIDEOPRESS) {
            iframeNode += `<script src="https://videopress.com/videopress-iframe.js"></script>`;
        }
        return iframeNode;
    }
}
exports.EmbedVideo = EmbedVideo;
